### PR TITLE
Implement ordered deployers by adding deploy.steps to config

### LINKED ISSUE #3905: Order of deployers doesn't match order listed in `skaffold.yaml`
### Expected Behavior

I have a `skaffold.yaml` file that contains the lines:

```yaml
apiVersion: skaffold/v2alpha3
kind: Config

metadata:
  name: discover

profiles:
  - name: local
    activation:
      - kubeContext: minikube
    build:
      local:
        useBuildkit: true
        push: false
      artifacts:
        - blah
    deploy:
      kubectl:
        manifests:
          - create-discover-app-namespace.json
      helm:
        releases:
          - name: dev
            chartPath: ./kubernetes/discover-app/
            etc: "etc"
```

Because `helm3` does not automatically create namespaces (anymore/yet), I have a `kubectl` deployer that is intended to create the namespace for the subsequent `helm` deployer. (I also have in mind some other things I&rsquo;d like to do with `kubectl` before `helm` runs - basically, I think there are other use cases besides creating namespaces.)

The `kubectl` deploy doesn&rsquo;t always run first, then the `helm` deploy fails. Following is a very trimmed debug output showing that.

I expect the `kubectl` deployer to run to completion before the `helm` deployer runs.

I have not at all investigated the `skaffold` source to see what changes could be made, sorry. If this issue is a low priority I may get to that and augment this issue with more information.

### Actual Behavior

See these logs; the `helm` deployer is running before the `kubectl` deployer and then nothing is working.

    (minikube:discover-app)➜  discover git:(acaird/AT-233-env-vars) ✗ skaffold -v debug run
    INFO[0000] Skaffold &{Version:v1.6.0-docs ConfigVersion:skaffold/v2beta1 GitVersion: GitCommit:b74e2f94f628b16a866abddc2ba8f05ce0bf956c GitTreeState:clean BuildDate:2020-03-25T00:09:12Z GoVersion:go1.14 Compiler:gc Platform:linux/amd64}
    DEBU[0000] config version (skaffold/v2alpha3) out of date: upgrading to latest (skaffold/v2beta1)
    INFO[0000] applying profile: local
    DEBU[0000] overlaying profile on config for field Build
     [...]
    INFO[0000] Using kubectl context: minikube
    DEBU[0000] Using builder: local
    DEBU[0000] Running command: [minikube docker-env --shell none]
    DEBU[0000] Command output: [DOCKER_TLS_VERIFY=1
    DOCKER_HOST=tcp://127.0.0.1:32769
    DOCKER_CERT_PATH=/home/acaird/.minikube/certs
    MINIKUBE_ACTIVE_DOCKERD=minikube
    ]
    DEBU[0000] setting Docker user agent to skaffold-v1.6.0-docs
    Generating tags...
     - discover-db-migrator -> DEBU[0000] Running command: [git describe --tags --always]
    DEBU[0000] Running command: [git describe --tags --always]
     [...]
    DEBU[0006] Running command: [helm version]
    INFO[0006] Deploying with helm v3.1.2 ...
    DEBU[0006] Executing template &{envTemplate 0xc0003fc900 0xc00003b7c0  } with environment map[COLORTERM:truecolor...
    DEBU[0006] Running command: [helm --kube-context minikube get all --namespace discover-app dev]
    Helm release dev not installed. Installing...
    DEBU[0006] EnvVarMap: map[DIGEST:discover-db-migrator:95cbaac2cf5f353ff87ca131d242f305f0a4b449b330b9faec0f63173fa7ed73...
    DEBU[0006] Executing template &{envTemplate 0xc00011f600 0xc000a560c0  } with environment map[COLORTERM:truecolor...
    DEBU[0006] Executing template &{envTemplate 0xc00062e000 0xc000a561c0  } with environment map[COLORTERM:truecolor...
     [...]
    DEBU[0006] Executing template &{envTemplate 0xc00062f200 0xc000a574c0  } with environment map[COLORTERM:truecolor...
    DEBU[0006] Executing template &{envTemplate 0xc00062f300 0xc000a57640  } with environment map[COLORTERM:truecolor...
    DEBU[0006] Running command: [helm --kube-context minikube install dev ./kubernetes/discover-app/ --namespace discover-app --set-string api.image.fullName=discover-web-server:797b046c95f1d9ad8a088fe841c62f5f385d8e4fed27a6cdb914317c7d414419 [...]
    coalesce.go:196: warning: cannot overwrite table with non table for nodeSelector (map[node_pool:association-node-pool])
    coalesce.go:196: warning: cannot overwrite table with non table for nodeSelector (map[node_pool:association-node-pool])
    Error: create: failed to create: namespaces "discover-app" not found
    DEBU[0007] Running command: [helm --kube-context minikube get all --namespace discover-app dev]
    WARN[0007] exit status 1
    DEBU[0007] getting client config for kubeContext: ``
    DEBU[0007] getting client config for kubeContext: ``
    DEBU[0007] Running command: [kubectl version --client -ojson]
    DEBU[0007] Command output: [{
      "clientVersion": {
        "major": "1",
        "minor": "17",
        "gitVersion": "v1.17.4",
        "gitCommit": "8d8aa39598534325ad77120c120a22b3a990b5ea",
        "gitTreeState": "clean",
        "buildDate": "2020-03-12T21:03:42Z",
        "goVersion": "go1.13.8",
        "compiler": "gc",
        "platform": "linux/amd64"
      }
    }
    ]
    DEBU[0007] Running command: [kubectl --context minikube create --dry-run -oyaml -f /home/acaird/git/discover/create-discover-app-namespace.json]
    DEBU[0007] Command output: [apiVersion: v1
    kind: Namespace
    metadata:
      labels:
        name: discover-app
      name: discover-app
    ]
    WARN[0007] image [discover-db-migrator] is not used by the deployment
     [...]
    DEBU[0007] manifests with tagged images: apiVersion: v1
    kind: Namespace
    metadata:
      labels:
        name: discover-app
      name: discover-app
    DEBU[0007] manifests with labels apiVersion: v1
    kind: Namespace
    metadata:
      labels:
        app.kubernetes.io/managed-by: skaffold-v1.6.0-docs
        name: discover-app
        skaffold.dev/builder: local
        skaffold.dev/cleanup: "true"
        skaffold.dev/deployer: kubectl
        skaffold.dev/docker-api-version: "1.40"
        skaffold.dev/run-id: 2bfe88b8-fe10-437f-8182-b82b55589ab2
        skaffold.dev/tag-policy: git-commit
        skaffold.dev/tail: "true"
      name: discover-app
    DEBU[0007] 1 manifests to deploy. 1 are updated or new
    DEBU[0007] Running command: [kubectl --context minikube apply -f -]
     - namespace/discover-app created
    INFO[0007] Deploy complete in 688.331603ms
    Waiting for deployments to stabilize
    DEBU[0007] getting client config for kubeContext: ``
    Deployments stabilized in 11.318892ms
    You can also run [skaffold run --tail] to get the logs

### Information

-   Skaffold Version:
    
        ~ skaffold version
        v1.6.0-docs
    
    and also:
    
        INFO[0000] Skaffold &{Version:v1.6.0-docs ConfigVersion:skaffold/v2beta1 GitVersion: GitCommit:b74e2f94f628b16a866abddc2ba8f05ce0bf956c GitTreeState:clean BuildDate:2020-03-25T00:09:12Z GoVersion:go1.14 Compiler:gc Platform:linux/amd64}

-   Operating System:
    
        ~ cat /etc/os-release
        NAME=Fedora
        VERSION="31 (Workstation Edition)"
        ID=fedora
        VERSION_ID=31
        VERSION_CODENAME=""
        PLATFORM_ID="platform:f31"
        PRETTY_NAME="Fedora 31 (Workstation Edition)"
        ANSI_COLOR="0;34"
        LOGO=fedora-logo-icon
        CPE_NAME="cpe:/o:fedoraproject:fedora:31"
        HOME_URL="https://fedoraproject.org/"
        DOCUMENTATION_URL="https://docs.fedoraproject.org/en-US/fedora/f31/system-administrators-guide/"
        SUPPORT_URL="https://fedoraproject.org/wiki/Communicating_and_getting_help"
        BUG_REPORT_URL="https://bugzilla.redhat.com/"
        REDHAT_BUGZILLA_PRODUCT="Fedora"
        REDHAT_BUGZILLA_PRODUCT_VERSION=31
        REDHAT_SUPPORT_PRODUCT="Fedora"
        REDHAT_SUPPORT_PRODUCT_VERSION=31
        PRIVACY_POLICY_URL="https://fedoraproject.org/wiki/Legal:PrivacyPolicy"
        VARIANT="Workstation Edition"
        VARIANT_ID=workstation
    
        ~ uname -a
        Linux primary 5.5.10-200.fc31.x86_64 #1 SMP Wed Mar 18 14:21:38 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
    
    See above for part of the `skaffold.yaml` file.

### Steps to Reproduce the Behavior

I think that any example of a `kubectl` deployer (maybe a slow one, like `create ns`) followed by a `helm` deployer should show this. TBH I haven&rsquo;t tried to create a minimal reproducible example, so let me know if that is needed.

### PR DESCRIPTION
**Pull request by MarlonGamez**
Fixes #3905 

**Description**
This change removes deploy.DeployType and adds deploy.steps, which is an array of DeployType, to the skaffold config.
This allows users to both specify the order of deployers, and allow for multiple of the same type of deployer to be run in sequence. 

**User facing changes**
The main changes are described above. Some config change examples can be seen here:
Before: 
```yaml
apiVersion: skaffold/v2beta8
kind: Config
build:
  artifacts:
  - image: skaffold-example
deploy:
  kubectl:
    manifests:
      - k8s-*
```
After:
```yaml
apiVersion: skaffold/v2beta9
kind: Config
build:
  artifacts:
  - image: skaffold-example
deploy:
  steps:
  - kubectl:
      manifests:
        - k8s-*
```

If a user wanted to have two kubectl deployers run in sequence, they could write a config like this:
```yaml
apiVersion: skaffold/v2beta9
kind: Config
build:
  artifacts:
    - image: leeroy-web
      context: leeroy-web
    - image: leeroy-app
      context: leeroy-app
deploy:
  steps:
  - kubectl:
      manifests:
      - leeroy-app/kubernetes/*
  - kubectl:
      manifests:
      - leeroy-web/kubernetes/*
portForward:
  - resourceType: deployment
    resourceName: leeroy-web
    port: 8080
    localPort: 9000
```

There are many files touched here, I apologize for that. Many of the changes are updating tests and their test data. If there is any possible way for me to split this into multiple PRs, please let me know.

### CODE REVIEW

File: examples/microservices/skaffold.yaml [hunk 1]
```diff
@@ -9,8 +9,8 @@ build:
 deploy:
   kubectl:
     manifests:
-      - leeroy-web/kubernetes/*
-      - leeroy-app/kubernetes/*
+    - leeroy-app/kubernetes/*
```
Comment (tejal29) [inline examples/microservices/skaffold.yaml:12]: is this intentional? this will be copied over during release.

**Review by tejal29 — COMMENTED** [review]
This looks really great!

Can you explain what changes are done to render when multiple deploy steps are present. 
Would render be called for each step and then the output can be consumed by kubectl?
e.g. the below works, 
```
skaffold render | kubectl apply -f - 
deployment.apps/leeroy-web created
service/leeroy-app created
```

Can users be still be able to do it?

Thanks
Tejal

File: examples/microservices/skaffold.yaml (hunk 1 above)
Comment (MarlonGamez) [reply examples/microservices/skaffold.yaml:12]: oh shoot, looks like I accidentally committed a change from when I was messing around with things. I'll fix that

**Comment by MarlonGamez** [comment]
@tejal29 Basically if there are multiple steps then the deployers are read into skaffold's preexisting mux deployer. When render is run with this deployer is calls Render on all of the deployers sequentially and concatenates the output with `"\n---\n"`

**Comment by MarlonGamez** [comment]
There seems to be a bug where this warning always shows when deploying
```
WARN[0001] image [gcr.io/marlongamez-main/leeroy-web] is not used by the deployment
WARN[0001] image [gcr.io/marlongamez-main/leeroy-app] is not used by the deployment
```

The deploy is still successful. Same thing when rendering. I'll look into this tomorrow

**Comment by yuwenma** [comment]
Thanks for letting me know this @MarlonGamez I guess my main question is that this approach seems to suggest users to mix the deployers. But this could cause more brittle to skaffold and causes more confusions. I've started the conversation to discuss this offline. Let's see if we can discuss it in the skaffold meeting.

File: pkg/skaffold/runner/new.go [hunk 2]
```diff
@@ -210,32 +210,32 @@ func getSyncer(cfg sync.Config) sync.Syncer {
...
-	}
+	for _, step := range cfg.Pipeline().Deploy.Steps {
+		if step.HelmDeploy != nil {
+			deployers = append(deployers, helm.NewDeployer(cfg, step, labels))
+		}
 
-	if d.KptDeploy != nil {
-		deployers = append(deployers, kpt.NewDeployer(cfg, labels))
-	}
+		if step.KptDeploy != nil {
```
Comment (tejal29) [inline pkg/skaffold/runner/new.go:220]: nit: tab present.

**Review by tejal29 — COMMENTED** [review]
Code change Looks good to me except for a nit.
Pending - Manual verification.

Should we change any existing integration test in testdata/ or add a new one to add multiple kubectl deploy steps? 
It would be great to add a example with a CRD and how steps can be used to order CRD creation.

Please go update the deploy doc https://skaffold.dev/docs/pipeline-stages/deployers/ to mention the `steps are performed in order`

**Comment by gsquared94** [comment]
> It would be great to add a example with a CRD and how steps can be used to order CRD creation.

+1, great opportunity to add an `ordered-deploy` example with a `CRD`

**Comment by MarlonGamez** [comment]
I'm closing this for now as there needs to be a larger discussion about skaffold's deploy phase as brought up by @nkubala

**Comment by gsquared94** [comment]
without supporting multiple deployers of each type I'm not sure it's possible to merge different deploy stanzas in Skaffold Modules (outlined in the modules [design](https://tinyurl.com/skaffold-modules)). WDYT @briandealwis @nkubala ?

**Pull request closed without merging**